#!/usr/bin/bash
#$ -cwd

# parameters 
CHR=$1
CHUNK_START=`printf "%.0f" $2`
CHUNK_END=`printf "%.0f" $3`
NE=$4  # NE=20000
XPAR=$5

# directories
ROOT_DIR=./
REF_DIR=/data/noether/datatmp-nobackup/tb2refDatabase/imputeRef/1000Gphase1/
PhaseRESULTS_DIR=${ROOT_DIR}3-phaseResults/
imputeRESULTS_DIR=${ROOT_DIR}5-imputeResults/

# executable (must be IMPUTE version 2.2.0 or later)
IMPUTE2_EXEC=/data/noether/tools/imputeScript/impute2
 
## MODIFY THE FOLLOWING THREE LINES TO ACCOMODATE OTHER PANELS
# reference data files
GENMAP_FILE=${REF_DIR}genetic_map_chr${CHR}_combined_b37.txt
HAPS_FILE=${REF_DIR}ALL_1000G_phase1integrated_v3_chr${CHR}_impute_macGT1.hap.gz
LEGEND_FILE=${REF_DIR}ALL_1000G_phase1integrated_v3_chr${CHR}_impute_macGT1.legend.gz
 

## THESE HAPLOTYPES WOULD BE GENERATED BY THE PREVIOUS SCRIPT
## SELECT ONE FROM METHOD-A AND METHOD-B BELOW
# METHOD-A: haplotypes from IMPUTE2 phasing run
# GWAS_HAPS_FILE=${PhaseRESULTS_DIR}gwas_data_chr${CHR}.pos${CHUNK_START}-${CHUNK_END}.phasing.impute2_haps
 
# METHOD-B: haplotypes from SHAPEIT phasing run
GWAS_HAPS_FILE=${PhaseRESULTS_DIR}chr${CHR}.haps
 
# main output file
OUTPUT_FILE=${imputeRESULTS_DIR}gwas_data_chr${CHR}.pos${CHUNK_START}-${CHUNK_END}.impute2
 
## impute genotypes from GWAS haplotypes
$IMPUTE2_EXEC \
   -iter 30 \
   -burnin 10 \
   -k_hap 500 \
   -use_prephased_g \
   $XPAR \
   -m $GENMAP_FILE \
   -h $HAPS_FILE \
   -l $LEGEND_FILE \
   -known_haps_g $GWAS_HAPS_FILE \
   -Ne $NE \
   -int $CHUNK_START $CHUNK_END \
   -buffer 1000 \
   -o $OUTPUT_FILE \
   -allow_large_regions \
   -seed 367946


